src/Microsoft.Maintainability.Analyzers/Core/AvoidUnusedPrivateFields.cs

```diff
@@ -32,7 +35,44 @@ public abstract class AvoidUnusedPrivateFieldsAnalyzer : DiagnosticAnalyzer
 
         public override void Initialize(AnalysisContext analysisContext)
         {
-            
+            analysisContext.RegisterCompilationStartAction(
+                (compilationContext) =>
+                {
+                    HashSet<IFieldSymbol> unreferencedPrivateFields = new HashSet<IFieldSymbol>();
+                    HashSet<IFieldSymbol> referencedPrivateFields = new HashSet<IFieldSymbol>();
+
+                    compilationContext.RegisterSymbolAction(
+                        (symbolContext) =>
+                        {
+                            IFieldSymbol field = (IFieldSymbol)symbolContext.Symbol;
+                            if (field.DeclaredAccessibility == Accessibility.Private && !referencedPrivateFields.Contains(field))
+                            {
+                                unreferencedPrivateFields.Add(field);
+                            }
+                        },
+                        SymbolKind.Field);
+
+                    compilationContext.RegisterOperationAction(
+                        (operationContext) =>
+                        {
+                            IFieldSymbol field = ((IFieldReferenceExpression)operationContext.Operation).Field;
+                            if (field.DeclaredAccessibility == Accessibility.Private)
+                            {
+                                referencedPrivateFields.Add(field);
+                                unreferencedPrivateFields.Remove(field);
+                            }
+                        },
+                        OperationKind.FieldReferenceExpression);
+
```

Comment:
Here's a question unrelated to this PR: does a symbol action always get invoked before an operation action in which the operation uses that symbol?